packages/multicast_dns/lib/src/resource_record.dart

```diff
@@ -0,0 +1,393 @@
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// A TXT (Text record) query.
+  ResourceRecordQuery.txt(
+    String name, {
+    bool isMulticast = true,
+  }) : this(
+          ResourceRecordType.txt,
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// Tye type of resource record - one of [ResourceRecordType]'s values.
+  final int resourceRecordType;
+
+  /// The Fully Qualified Domain Name associated with the request.
+  final String name;
+
+  /// The [QuestionType], i.e. multicast or unicast.
+  final int questionType;
+
+  /// Convenience accessor to determine whether the question type is multicast.
+  bool get isMulticast => questionType == QuestionType.multicast;
+
+  /// Convenience accessor to determine whether the question type is unicast.
+  bool get isUnicast => questionType == QuestionType.unicast;
+
+  /// Encodes this query to the raw wire format.
+  List<int> encode() {
+    return encodeMDnsQuery(
+      name,
+      type: resourceRecordType,
+      multicast: isMulticast,
+    );
+  }
+
+  @override
+  int get hashCode =>
+      _hashValues(<int>[resourceRecordType, name.hashCode, questionType]);
+
+  @override
+  bool operator ==(Object other) {
+    return other is ResourceRecordQuery &&
+        other.resourceRecordType == resourceRecordType &&
+        other.name == name &&
+        other.questionType == questionType;
+  }
+
+  @override
+  String toString() =>
+      '$runtimeType{$name, type: ${ResourceRecordType.toDebugString(resourceRecordType)}, isMulticast: $isMulticast}';
+}
+
+/// Base implementation of DNS resource records (RRs).
+abstract class ResourceRecord {
+  /// Creates a new ResourceRecord.
+  const ResourceRecord(this.resourceRecordType, this.name, this.validUntil)
+      : assert(name != null);
+
+  /// The FQDN for this record.
+  final String name;
+
+  /// The epoch time at which point this record is valid for in the cache.
+  final int validUntil;
+
+  /// The raw resource record value.  See [ResourceRecordType] for supported values.
+  final int resourceRecordType;
+
+  String get _additionalInfo;
+
+  @override
+  String toString() =>
+      '$runtimeType{$name, validUntil: ${DateTime.fromMillisecondsSinceEpoch(validUntil ?? 0)}, $_additionalInfo}';
+
+  @override
+  bool operator ==(Object other) {
+    return other.runtimeType == runtimeType && _equals(other);
+  }
+
+  @protected
+  bool _equals(ResourceRecord other) {
+    return other.name == name &&
+        other.validUntil == validUntil &&
+        other.resourceRecordType == resourceRecordType;
+  }
+
+  @override
+  int get hashCode {
+    return _hashValues(<int>[
+      name.hashCode,
+      validUntil.hashCode,
+      resourceRecordType.hashCode,
+      _hashCode,
+    ]);
+  }
+
+  @protected
+  int get _hashCode;
+
+  /// Low level method for encoding this record into an mDNS packet.
```

Comment:
If I'm going to subclass this, what would I have to do here? What are valid return values?